javascript/ql/src/experimental/Security/CWE-020/PostMessageNoOriginCheck.ql

```diff
@@ -0,0 +1,71 @@
+/**
+ * @name Missing `MessageEvent.origin` verification in `postMessage` handlers
+ * @description Missing the `MessageEvent.origin` verification in `postMessage` handlers, allows any windows to send arbitrary data to the `MessageEvent` listener.
+ * This could lead to unexpected behaviour, especially when `MessageEvent.data` is used in an unsafe way.
+ * @kind problem
+ * @problem.severity warning
+ * @precision high
+ * @id js/missing-postmessageorigin-verification
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-20
+ */
+
+import javascript
+import semmle.javascript.security.dataflow.DOM
+
+/**
+ * A method call for the insecure functions used to verify the `MessageEvent.origin`.
+ */
+class InsufficientOriginChecks extends DataFlow::MethodCallNode {
+  InsufficientOriginChecks() {
+    exists(string name | name = getMethodName() |
+      name = "indexOf" or
+      name = "includes" or
+      name = "endsWith" or
+      name = "startsWith" or
+      name = "lastIndexOf"
+    )
+  }
+}
+
+/**
+ * A function handler for the `MessageEvent`.
+ */
+class PostMessageHandler extends DataFlow::FunctionNode {
+  PostMessageHandler() { exists(PostMessageEventHandler handler | this.getFunction() = handler) }
+}
+
+/**
+ * The `MessageEvent` parameter received by the handler
+ */
+class PostMessageEvent extends DataFlow::SourceNode {
+  PostMessageEvent() { exists(PostMessageHandler handler | this = handler.getParameter(0)) }
+
+  VarAccess event;
+  EqualityTest astNode;
+
+  /**
+   * Holds if an access on `MessageEvent.origin` is in an `EqualityTest` and there is no call of an insufficient verification method on `MessageEvent.origin`
+   */
+  predicate hasOriginChecked() {
+    exists(string prop | prop = "origin" or prop = "source" |
+      astNode.getAnOperand().(PropAccess).accesses(event, prop) and
+      event.mayReferToParameter*(this.asExpr()) and
+      not this.hasOriginInsufficientlyChecked()
+    )
```

Comment:
Commenting on the QL. 
I think this is what you meant to write: "the origin or source property of the event is used in an equality test, ~and it is not insufficiently checked~".

```suggestion
    exists(EqualityTest test |
      this.getAPropertyRead(["origin", "source"]).flowsToExpr(test.getAnOperand())
   )
```

Reply:
But what I really think you want is:

```ql
exists(MembershipCandidate cand | this.flowsTo(cand))
```
This class handles many more cases than just `==` and its friends.

See:
https://github.com/github/codeql/blob/d80a033bed070c98ce91b74986098297e5e11ab1/javascript/ql/src/semmle/javascript/MembershipCandidates.qll#L12

(yes, we should update `PostMessageEventSanitizer` similarly)

Reply:
Yes, this is exactly what I wanted to achieve. Thanks for pointing this out!